### PR TITLE
New env var to choose artifcats storage root when tracking on db

### LINKED ISSUE #4077: [BUG] No way to configure artifact store path when tracking on SQLAlchemy
## Willingness to contribute

- [X ] Yes. I can contribute this feature independently.

## Proposal Summary

By default, artifacts are stored to `./mlruns` (hard coded constant: DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH). The idea is to give the possibility to change this behaviour by setting an env var.

The same mechanism exists but not generalized:

```
_TRACKING_DIR_ENV_VAR = "MLFLOW_TRACKING_DIR"

def _default_root_dir():
    return get_env(_TRACKING_DIR_ENV_VAR) or os.path.abspath(DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH)
```

### PR DESCRIPTION
**Pull request by alikefia**
## What changes are proposed in this pull request?

New env var to configure artifacts storage root when tracking to db via SQLAlchemy (#4077).
- When tracking on files, this path is defined via `MLFLOW_TRACKING_URI`
- When tracking to DB, MLFLOW_TRACKING_URI is used to setup the connection to Database and the file storage for artifacts is hard coded to `DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH=./mlruns`

## How is this patch tested?

N/A

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Gives the possibility to configure the root directory where artifacts are stored when tracking on DB.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [X] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Review by tomzx — COMMENTED** [review]
Change makes sense. Can you add some tests in `tests/tracking/_tracking_service/test_utils.py`?

**Comment by alikefia** [comment]
Done.
Picked the fix for list_experiments (SQLAlchemy v1.4) to run the tests locally.

File: mlflow/tracking/_tracking_service/utils.py [hunk 1]
```diff
@@ -113,7 +114,9 @@ def _get_sqlalchemy_store(store_uri, artifact_uri):
     from mlflow.store.tracking.sqlalchemy_store import SqlAlchemyStore
 
     if artifact_uri is None:
-        artifact_uri = DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH
+        artifact_uri = (
+            os.environ.get(_ARTIFACT_ROOT_ENV_VAR, None) or DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH
```
Comment (tomzx) [inline mlflow/tracking/_tracking_service/utils.py:118]: Is there a reason you preferred this over
```suggestion
            os.environ.get(_ARTIFACT_ROOT_ENV_VAR, DEFAULT_LOCAL_FILE_AND_ARTIFACT_PATH)
```
While I understand an empty/unset `MLFLOW_ARTIFACT_ROOT`  would be replaced by the default value, is there another reason?

From the code above (`get_tracking_uri`), it appears that if the `_TRACKING_URI_ENV_VAR` environment variable is set but empty, we would still use its empty value. Thus the two behavior are different which might not be what we want.

File: tests/tracking/_tracking_service/test_utils.py [hunk 2]
```diff
@@ -155,6 +168,31 @@ def test_get_store_sqlalchemy_store_with_artifact_uri(tmp_wkdir, db_type):
     mock_create_engine.assert_called_once_with(uri, pool_pre_ping=True)
 
 
+def test_get_store_sqlalchemy_store_with_artifact_root():
```
Comment (tomzx) [inline tests/tracking/_tracking_service/test_utils.py:171]: Would it be possible to convert the 3 tests (+ this one) that share almost the same code into a parameterized test?

Comment (alikefia) [reply tests/tracking/_tracking_service/test_utils.py:171]: No reason indeed, totally agree

File: mlflow/tracking/_tracking_service/utils.py (hunk 1 above)
Comment (alikefia) [reply mlflow/tracking/_tracking_service/utils.py]: No reason indeed, agree !

File: tests/tracking/_tracking_service/test_utils.py (hunk 2 above)
Comment (alikefia) [reply tests/tracking/_tracking_service/test_utils.py]: Done

**Comment by shachibista** [comment]
Any updates on this?

**Comment by andreweskeclarke** [comment]
Agreed, any updates here? Would love to be able to set the artifacts store like I can set the mlruns.db path. In particular I am trying to add integration tests and this is unnecessarily creating an ./mlruns directory that I can't put in my `tmp_path` location.

**Comment by andreweskeclarke** [comment]
@tomzx (see above comment)

**Pull request closed without merging**